File: revisiting_ctci/chapter4/MinimalTree_4_2.go
/*
Given a sorted array with unique integer elements
write an algorithm to create a binary search tree with
minimal height
*/
package main

import (
	"fmt"
)

type BNode struct{
	Data int
	Left *BNode
	Right *BNode
}



//take a sorted list and construct a binary search tree
func BuildBalancedTree(s_list []int) BNode{
	if len(s_list) == 0{
		return BNode{}
	}
	middle := int(len(s_list)/2)

	pos := BNode{Data : s_list[middle]}


	if len(s_list[:middle]) > 0{
		left := BuildBalancedTree(s_list[:middle])
		pos.Left = &left	
	} else {
		pos.Left = nil
	}
	
	if len(s_list[middle+1:]) > 0 {
		right := BuildBalancedTree(s_list[middle+1:])
		pos.Right = &right
	} else{
		pos.Right = nil
	}

	return pos
}


func main(){
	sorted_array := []int{1,3,5,7,9,11,13}

	ans := BuildBalancedTree(sorted_array)

	fmt.Println(ans.Data == 7)
	left_ans := *ans.Left
	fmt.Println(left_ans.Data == 3)
	right_left_ans := *left_ans.Right
	fmt.Println(right_left_ans.Data == 5)
}

File: revisiting_ctci/chapter4/MinimalTree_4_2.py
"""
Given a sorted array with unique integer elements
write an algorithm to create a binary dearch tree with
minimal height
"""
from graph import Node, Graph

class BalancedNode():
	def __init__(self, data=None):
		#this stores the data found at the given node
		self.__data = data
		#this stores the adjacent edges in the graph, that can
		#be reached from the given position
		self.__left = None
		self.__right = None

	def LeftAdd(self, data):
		"""take a new node and add it to the list of adjacent positions """
		self.__left = data

	def RightAdd(self, data):
		"""take a new node and add it to the list of adjacent positions """
		self.__right = data

	def ChangeVal(self, value):
		""" change the value stored for the given node """
		self.__data = value

	@property
	def Data(self):
		return self.__data

	@property
	def Left(self):
		return self.__left

	@property
	def Right(self):
		return self.__right

	def __repr__(self):
		return f'{self.Data}'


def BuildBalancedGraph(sorted_array):
	if len(sorted_array) == 0:
		return []

	middle = int(len(sorted_array)/2)
	pos = BalancedNode(sorted_array[middle])

	left = BuildBalancedGraph(sorted_array[:middle])
	right = BuildBalancedGraph(sorted_array[middle+1:])

	pos.LeftAdd(left)
	pos.RightAdd(right)

	return pos

if __name__ == '__main__':


	sorted_array = [1,3,5,7,9,11,13]

	ans = BuildBalancedGraph(sorted_array)

	ans.Data
	x = ans.Left
	y = x.Left
	y.Left